filters/include/pcl/filters/impl/conditional_removal.hpp

```diff
@@ -525,60 +525,33 @@ pcl::PointDataAtOffset<PointT>::compare (const PointT& p, const double& val)
   
   const std::uint8_t* pt_data = reinterpret_cast<const std::uint8_t*> (&p);
 
-  switch (datatype_) 
-  {
-    case pcl::PCLPointField::INT8 :
-    {
-      std::int8_t pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (std::int8_t));
-      return (pt_val > static_cast<std::int8_t>(val)) - (pt_val < static_cast<std::int8_t> (val));
-    }
-    case pcl::PCLPointField::UINT8 :
-    {
-      std::uint8_t pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (std::uint8_t));
-      return (pt_val > static_cast<std::uint8_t>(val)) - (pt_val < static_cast<std::uint8_t> (val));
-    }
-    case pcl::PCLPointField::INT16 :
-    {
-      std::int16_t pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (std::int16_t));
-      return (pt_val > static_cast<std::int16_t>(val)) - (pt_val < static_cast<std::int16_t> (val));
-    }
-    case pcl::PCLPointField::UINT16 :
-    {
-      std::uint16_t pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (std::uint16_t));
-      return (pt_val > static_cast<std::uint16_t> (val)) - (pt_val < static_cast<std::uint16_t> (val));
-    }
-    case pcl::PCLPointField::INT32 :
-    {
-      std::int32_t pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (std::int32_t));
-      return (pt_val > static_cast<std::int32_t> (val)) - (pt_val < static_cast<std::int32_t> (val));
-    }
-    case pcl::PCLPointField::UINT32 :
-    {
-      std::uint32_t pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (std::uint32_t));
-      return (pt_val > static_cast<std::uint32_t> (val)) - (pt_val < static_cast<std::uint32_t> (val));
-    }
-    case pcl::PCLPointField::FLOAT32 :
-    {
-      float pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (float));
-      return (pt_val > static_cast<float> (val)) - (pt_val < static_cast<float> (val));
-    }
-    case pcl::PCLPointField::FLOAT64 :
-    {
-      double pt_val;
-      memcpy (&pt_val, pt_data + this->offset_, sizeof (double));
-      return (pt_val > val) - (pt_val < val);
-    }
-    default : 
+#define SAFE_COMPARE(CASE_LABEL)                                                       \
+  case CASE_LABEL: {                                                                   \
+    pcl::traits::asType_t<CASE_LABEL> pt_val;                                          \
+    memcpy(&pt_val, pt_data + this->offset_, sizeof(pt_val));                          \
+    return (pt_val > static_cast<pcl::traits::asType_t<CASE_LABEL>>(val)) -            \
+           (pt_val < static_cast<pcl::traits::asType_t<CASE_LABEL>>(val));             \
+  }
+
+  switch (datatype_)
+  {
+    SAFE_COMPARE(pcl::PCLPointField::BOOL)
```

Comment:
true > false should be sufficient, right?

Reply:
IMO `true` and `false` should not have any greater/less relationship, only equal/not equal. Might be best to create a new subclass to `ComparisonBase`, something like `BoolComparison`? There `compare_val` could be a `bool`.

Reply:
Strong ordering is well defined for booleans. I have a feeling I'm missing something here

Reply:
PTAL

Reply:
> Strong ordering is well defined for booleans

Could you give a link for that? I didn't really find anything on that topic.
I assume `true>false` is because `1>0`, but just from a logical point of view, it makes no sense that `true` is bigger than `false` - you cannot order them.
And `val` (a `double`) could be `0.0` for `false` and  `1.0` for `true`, but if feel like there are potentially a lot of hidden bugs. Representing a boolean with a double does not make as much sense as representing an actual number with a double.

Reply:
https://stackoverflow.com/questions/11951729

I'll take another pass at the logic though

Reply:
The main logic of the static_cast is that only non-zero values are true, and they are all true. So,
```cpp
bool a = -0.001;
bool b = 9999.222;

// these are all true
assert(a == b);
assert(a == true);
assert(false < a);
assert(false < b);
```